File: app/src/main/java/com/settlex/android/presentation/dashboard/services/adapter/ServicesAdapter.java
package com.settlex.android.presentation.dashboard.services.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.settlex.android.R;
import com.settlex.android.presentation.dashboard.services.model.ServiceUiModel;

import java.util.List;

/**
 * RecyclerView adapter for displaying services in a grid/list format
 */
public class ServicesAdapter extends RecyclerView.Adapter<ServicesAdapter.ServiceViewHolder> {
    private final List<ServiceUiModel> service;
    private final onItemClickedListener listener;
    private final boolean useAllService; // true = item_service_all

    public interface onItemClickedListener {
        void onServiceClick(ServiceUiModel service);
    }

    public ServicesAdapter(boolean useAllService, List<ServiceUiModel> service, onItemClickedListener listener) {
        this.useAllService = useAllService;
        this.service = service;
        this.listener = listener;
    }

    @NonNull
    @Override
    public ServiceViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        int layoutId = (!useAllService) ? R.layout.item_service : R.layout.item_service_all;
        View view = LayoutInflater.from(parent.getContext()).inflate(layoutId, parent, false);
        return new ServiceViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ServiceViewHolder holder, int position) {
        ServiceUiModel service = this.service.get(position);

        // Bind service data to views
        holder.name.setText(service.getName());
        holder.icon.setImageResource(service.getIconResId());

        if (!(service.getCashbackPercentage() < 1)) {
            String CASHBACK = "up to " + service.getCashbackPercentage() + "%";
            holder.badge.setText(CASHBACK);
            holder.badge.setVisibility(View.VISIBLE);

        } else if (service.getLabel() != null && !service.getLabel().isEmpty()) {
            holder.badge.setText(service.getLabel());
        } else {
            holder.badge.setVisibility(View.GONE);
        }
        holder.itemView.setOnClickListener(view -> listener.onServiceClick(this.service.get(position)));
    }

    @Override
    public int getItemCount() {
        return service.size();
    }

    public static class ServiceViewHolder extends RecyclerView.ViewHolder {
        ImageView icon;
        TextView name, badge;

        ServiceViewHolder(@NonNull View itemView) {
            super(itemView);
            icon = itemView.findViewById(R.id.icon);
            name = itemView.findViewById(R.id.name);
            badge = itemView.findViewById(R.id.badge);
        }
    }
}

File: app/src/main/java/com/settlex/android/presentation/dashboard/services/adapter/ServicesAdapter.kt
package com.settlex.android.presentation.dashboard.services.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.settlex.android.R
import com.settlex.android.presentation.common.extensions.gone
import com.settlex.android.presentation.common.extensions.show
import com.settlex.android.presentation.dashboard.services.adapter.ServicesAdapter.ServiceViewHolder
import com.settlex.android.presentation.dashboard.services.model.ServiceUiModel

/**
 * RecyclerView adapter for displaying services in a grid/list format
 */
class ServicesAdapter(
    private val isHomeDashboard: Boolean,
    private val service: List<ServiceUiModel>,
    private val listener: OnServiceClickedListener
) : RecyclerView.Adapter<ServiceViewHolder>() {

    fun interface OnServiceClickedListener {
        fun onServiceClick(service: ServiceUiModel)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ServiceViewHolder {
        val layoutId = when {
            isHomeDashboard -> R.layout.item_service_home
            else -> R.layout.item_service_full
        }

        val view = LayoutInflater.from(parent.context).inflate(layoutId, parent, false)
        return ServiceViewHolder(view)
    }

    override fun onBindViewHolder(holder: ServiceViewHolder, position: Int) = with(holder) {
        val service = service[position]

        // Bind service data to views
        tvServiceName.text = service.name
        ivServiceIcon.setImageResource(service.iconResId)

        when {
            service.cashbackPercentage > 0 -> {
                "up to ${service.cashbackPercentage} %".also { holder.tvServiceBadge.text = it }
                tvServiceBadge.show()
            }

            !service.label.isNullOrEmpty() -> tvServiceBadge.text = service.label

            else -> tvServiceBadge.gone()
        }

        itemView.setOnClickListener { listener.onServiceClick(service) }
    }

    override fun getItemCount(): Int {
        return service.size
    }

    class ServiceViewHolder internal constructor(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var ivServiceIcon: ImageView = itemView.findViewById(R.id.iv_service_icon)
        var tvServiceName: TextView = itemView.findViewById(R.id.tv_service_name)
        var tvServiceBadge: TextView = itemView.findViewById(R.id.tv_service_badge)
    }
}